File: Backend/services/reportService.js
const Booking = require("../models/bookingModel");
const Slot = require("../models/slotModel");

/**
 * Generate comprehensive usage report
 * @param {Date} startDate - Start date for report
 * @param {Date} endDate - End date for report
 * @param {Object} filters - Additional filters (slotId, userType, city, area, address)
 * @returns {Object} Report data with metrics and breakdowns
 */
const generateUsageReport = async (startDate, endDate, filters = {}) => {
    try {
        // Build query
        const query = {
            createdAt: {
                $gte: new Date(startDate),
                $lte: new Date(endDate)
            }
        };

        // Apply filters
        if (filters.slotId) {
            query.slot = filters.slotId;
        }

        // Fetch bookings with populated data
        const bookings = await Booking.find(query)
            .populate({
                path: "user",
                select: "name email role"
            })
            .populate({
                path: "slot",
                select: "slotNumber city area address placeType"
            })
            .sort({ createdAt: -1 });

        // Filter by user type if specified
        let filteredBookings = bookings;
        if (filters.userType && filters.userType !== 'all') {
            filteredBookings = bookings.filter(b => b.user?.role === filters.userType);
        }

        // Filter by location if specified
        if (filters.city) {
            filteredBookings = filteredBookings.filter(b => b.slot?.city === filters.city);
        }
        if (filters.area) {
            filteredBookings = filteredBookings.filter(b => b.slot?.area === filters.area);
        }
        if (filters.address) {
            filteredBookings = filteredBookings.filter(b => b.slot?.address === filters.address);
        }

        // Calculate metrics
        const totalBookings = filteredBookings.length;
        const activeBookings = filteredBookings.filter(b => b.parkingStatus === 'CHECKED_IN').length;
        const completedBookings = filteredBookings.filter(b => b.status === 'COMPLETED').length;
        const cancelledBookings = filteredBookings.filter(b => b.status === 'CANCELLED').length;

        // Revenue calculations
        const totalRevenue = filteredBookings.reduce((sum, b) => sum + (b.payment?.amount || 0), 0);
        const averageRevenue = totalBookings > 0 ? totalRevenue / totalBookings : 0;
        const paidBookings = filteredBookings.filter(b => b.payment?.status === 'completed').length;

        // Duration calculations
        const averageDuration = calculateAverageDuration(filteredBookings);

        // Peak hours analysis
        const peakHours = calculatePeakHours(filteredBookings);

        // Segmentation
        const byUserType = segmentByUserType(filteredBookings);
        const bySlot = segmentBySlot(filteredBookings);
        const byLocation = segmentByLocation(filteredBookings);
        const byStatus = segmentByStatus(filteredBookings);

        // Time series data (for charts)
        const timeSeriesData = generateTimeSeriesData(filteredBookings, startDate, endDate);

        return {
            summary: {
                totalBookings,
                activeBookings,
                completedBookings,
                cancelledBookings,
                totalRevenue: parseFloat(totalRevenue.toFixed(2)),
                averageRevenue: parseFloat(averageRevenue.toFixed(2)),
                paidBookings,
                averageDuration: parseFloat(averageDuration.toFixed(2))
            },
            peakHours,
            segmentation: {
                byUserType,
                bySlot,
                byLocation,
                byStatus
            },
            timeSeries: timeSeriesData,
            dateRange: {
                start: startDate,
                end: endDate
            },
            filters: filters
        };
    } catch (error) {
        console.error("Error generating usage report:", error);
        throw new Error("Failed to generate usage report");
    }
};

/**
 * Calculate peak hours from bookings
 */
const calculatePeakHours = (bookings) => {
    const hourCounts = {};

    bookings.forEach(booking => {
        const hour = new Date(booking.startTime).getHours();
        hourCounts[hour] = (hourCounts[hour] || 0) + 1;
    });

    // Convert to array and sort by count
    const hourArray = Object.entries(hourCounts).map(([hour, count]) => ({
        hour: parseInt(hour),
        count,
        timeLabel: `${hour}:00 - ${hour}:59`
    })).sort((a, b) => b.count - a.count);

    return {
        peakHour: hourArray[0]?.timeLabel || 'N/A',
        peakHourCount: hourArray[0]?.count || 0,
        distribution: hourArray
    };
};

/**
 * Calculate average parking duration in minutes
 */
const calculateAverageDuration = (bookings) => {
    const completedBookings = bookings.filter(b =>
        b.actualEntryTime && b.actualExitTime
    );

    if (completedBookings.length === 0) {
        // Fallback to scheduled duration
        const scheduledDurations = bookings.map(b => {
            const start = new Date(b.startTime);
            const end = new Date(b.endTime);
            return (end - start) / (1000 * 60); // Convert to minutes
        });

        return scheduledDurations.length > 0
            ? scheduledDurations.reduce((sum, d) => sum + d, 0) / scheduledDurations.length
            : 0;
    }

    const totalDuration = completedBookings.reduce((sum, b) => {
        return sum + (b.actualDuration || 0);
    }, 0);

    return totalDuration / completedBookings.length;
};

/**
 * Segment bookings by user type
 */
const segmentByUserType = (bookings) => {
    const segments = {
        admin: 0,
        user: 0,
        unknown: 0
    };

    bookings.forEach(booking => {
        const role = booking.user?.role || 'unknown';
        segments[role] = (segments[role] || 0) + 1;
    });

    return segments;
};

/**
 * Segment bookings by slot
 */
const segmentBySlot = (bookings) => {
    const slotMap = {};

    bookings.forEach(booking => {
        const slotNumber = booking.slot?.slotNumber || 'Unknown';
        if (!slotMap[slotNumber]) {
            slotMap[slotNumber] = {
                slotNumber,
                slotId: booking.slot?._id,
                count: 0,
                revenue: 0
            };
        }
        slotMap[slotNumber].count++;
        slotMap[slotNumber].revenue += booking.payment?.amount || 0;
    });

    // Convert to array and sort by count
    return Object.values(slotMap)
        .sort((a, b) => b.count - a.count)
        .slice(0, 10); // Top 10 slots
};

/**
 * Segment bookings by location
 */
const segmentByLocation = (bookings) => {
    const locationMap = {};

    bookings.forEach(booking => {
        const location = booking.slot?.address || 'Unknown';
        const city = booking.slot?.city || 'Unknown';
        const area = booking.slot?.area || 'Unknown';

        const key = `${city} - ${area} - ${location}`;

        if (!locationMap[key]) {
            locationMap[key] = {
                city,
                area,
                address: location,
                count: 0,
                revenue: 0
            };
        }
        locationMap[key].count++;
        locationMap[key].revenue += booking.payment?.amount || 0;
    });

    return Object.values(locationMap).sort((a, b) => b.count - a.count);
};

/**
 * Segment bookings by status
 */
const segmentByStatus = (bookings) => {
    const statusMap = {
        BOOKED: 0,
        COMPLETED: 0,
        CANCELLED: 0
    };

    bookings.forEach(booking => {
        const status = booking.status || 'BOOKED';
        statusMap[status] = (statusMap[status] || 0) + 1;
    });

    return statusMap;
};

/**
 * Generate time series data for charts
 */
const generateTimeSeriesData = (bookings, startDate, endDate) => {
    const start = new Date(startDate);
    const end = new Date(endDate);
    const daysDiff = Math.ceil((end - start) / (1000 * 60 * 60 * 24));

    const timeSeriesMap = {};

    // Initialize all dates in range
    for (let i = 0; i <= daysDiff; i++) {
        const date = new Date(start);
        date.setDate(date.getDate() + i);
        const dateKey = date.toISOString().split('T')[0];
        timeSeriesMap[dateKey] = {
            date: dateKey,
            bookings: 0,
            revenue: 0,
            completed: 0,
            cancelled: 0
        };
    }

    // Populate with booking data
    bookings.forEach(booking => {
        const dateKey = new Date(booking.createdAt).toISOString().split('T')[0];
        if (timeSeriesMap[dateKey]) {
            timeSeriesMap[dateKey].bookings++;
            timeSeriesMap[dateKey].revenue += booking.payment?.amount || 0;
            if (booking.status === 'COMPLETED') timeSeriesMap[dateKey].completed++;
            if (booking.status === 'CANCELLED') timeSeriesMap[dateKey].cancelled++;
        }
    });

    return Object.values(timeSeriesMap).sort((a, b) =>
        new Date(a.date) - new Date(b.date)
    );
};

/**
 * Generate CSV data from report
 */
const generateCSVData = (reportData, bookings) => {
    // Prepare CSV rows
    const csvRows = bookings.map(booking => ({
        'Booking ID': booking._id,
        'Date': new Date(booking.createdAt).toLocaleDateString(),
        'Time': new Date(booking.createdAt).toLocaleTimeString(),
        'User Name': booking.user?.name || 'N/A',
        'User Email': booking.user?.email || 'N/A',
        'User Type': booking.user?.role || 'N/A',
        'Slot Number': booking.slot?.slotNumber || 'N/A',
        'City': booking.slot?.city || 'N/A',
        'Area': booking.slot?.area || 'N/A',
        'Address': booking.slot?.address || 'N/A',
        'Place Type': booking.slot?.placeType || 'N/A',
        'Vehicle Number': booking.vehicleNumber,
        'Start Time': new Date(booking.startTime).toLocaleString(),
        'End Time': new Date(booking.endTime).toLocaleString(),
        'Parking Status': booking.parkingStatus,
        'Booking Status': booking.status,
        'Duration (min)': booking.actualDuration || 'N/A',
        'Amount': booking.payment?.amount || 0,
        'Payment Method': booking.payment?.method || 'N/A',
        'Payment Status': booking.payment?.status || 'N/A',
        'Transaction ID': booking.payment?.transactionId || 'N/A'
    }));

    return csvRows;
};

/**
 * Get personal statistics for a user
 */
const getPersonalStats = async (userId) => {
    try {
        const bookings = await Booking.find({ user: userId })
            .populate('slot', 'slotNumber city area address placeType')
            .sort({ createdAt: -1 });

        const totalBookings = bookings.length;
        const totalSpent = bookings.reduce((sum, b) => sum + (b.payment?.amount || 0), 0);
        const averageDuration = calculateAverageDuration(bookings);

        // Find favorite location
        const locationCounts = {};
        bookings.forEach(b => {
            const loc = b.slot?.address || 'Unknown';
            locationCounts[loc] = (locationCounts[loc] || 0) + 1;
        });
        const favoriteLocation = Object.entries(locationCounts)
            .sort((a, b) => b[1] - a[1])[0]?.[0] || 'N/A';

        // Last 30 days data
        const thirtyDaysAgo = new Date();
        thirtyDaysAgo.setDate(thirtyDaysAgo.getDate() - 30);
        const recentBookings = bookings.filter(b =>
            new Date(b.createdAt) >= thirtyDaysAgo
        );

        const timeSeriesData = generateTimeSeriesData(
            recentBookings,
            thirtyDaysAgo,
            new Date()
        );

        // Most used time slots
        const peakHours = calculatePeakHours(bookings);

        return {
            summary: {
                totalBookings,
                totalSpent: parseFloat(totalSpent.toFixed(2)),
                averageDuration: parseFloat(averageDuration.toFixed(2)),
                favoriteLocation,
                activeBookings: bookings.filter(b => b.parkingStatus === 'CHECKED_IN').length,
                completedBookings: bookings.filter(b => b.status === 'COMPLETED').length
            },
            timeSeries: timeSeriesData,
            peakHours: peakHours.distribution.slice(0, 5), // Top 5 hours
            recentBookings: recentBookings.slice(0, 10).map(b => ({
                id: b._id,
                slotNumber: b.slot?.slotNumber,
                date: b.createdAt,
                amount: b.payment?.amount || 0,
                status: b.status
            }))
        };
    } catch (error) {
        console.error("Error getting personal stats:", error);
        throw new Error("Failed to get personal statistics");
    }
};

module.exports = {
    generateUsageReport,
    calculatePeakHours,
    calculateAverageDuration,
    segmentByUserType,
    segmentBySlot,
    segmentByLocation,
    generateCSVData,
    getPersonalStats
};

File: validator_checker/src/main/java/com/parking/validator/service/ReportService.java
package com.parking.validator.service;

import com.parking.validator.model.Booking;
import com.parking.validator.repository.BookingRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class ReportService {

        @Autowired
        private BookingRepository bookingRepository;

        public Map<String, Object> generateUsageReport(LocalDateTime start, LocalDateTime end,
                        Map<String, String> filters) {
                List<Booking> bookings = bookingRepository.findAll().stream()
                                .filter(b -> b.getCreatedAt() != null && !b.getCreatedAt().isBefore(start)
                                                && !b.getCreatedAt().isAfter(end))
                                .collect(Collectors.toList());

                // Apply filters
                if (filters.get("slotId") != null) {
                        String slotId = filters.get("slotId");
                        bookings = bookings.stream()
                                        .filter(b -> b.getSlot() != null && b.getSlot().getId().equals(slotId))
                                        .collect(Collectors.toList());
                }
                if (filters.get("city") != null) {
                        String city = filters.get("city");
                        bookings = bookings.stream()
                                        .filter(b -> b.getSlot() != null && city.equals(b.getSlot().getCity()))
                                        .collect(Collectors.toList());
                }
                if (filters.get("area") != null) {
                        String area = filters.get("area");
                        bookings = bookings.stream()
                                        .filter(b -> b.getSlot() != null && area.equals(b.getSlot().getArea()))
                                        .collect(Collectors.toList());
                }

                Map<String, Object> report = new HashMap<>();
                Map<String, Object> summary = new HashMap<>();
                summary.put("totalBookings", bookings.size());
                summary.put("activeBookings",
                                bookings.stream().filter(b -> b.getParkingStatus() == Booking.ParkingStatus.CHECKED_IN)
                                                .count());
                summary.put("completedBookings",
                                bookings.stream().filter(b -> b.getStatus() == Booking.BookingStatus.COMPLETED)
                                                .count());
                summary.put("totalRevenue", bookings.stream()
                                .filter(b -> b.getPayment() != null
                                                && b.getPayment().getStatus() == Booking.PaymentStatus.completed)
                                .mapToDouble(b -> b.getPayment().getAmount() != null ? b.getPayment().getAmount() : 0.0)
                                .sum());

                double avgDuration = bookings.stream()
                                .filter(b -> b.getActualDuration() != null && b.getActualDuration() > 0)
                                .mapToInt(b -> b.getActualDuration())
                                .average().orElse(0.0);
                summary.put("averageDuration", Math.round(avgDuration));

                report.put("summary", summary);
                report.put("peakHours", calculatePeakHours(bookings));
                report.put("segmentation", calculateSegmentation(bookings));
                report.put("timeSeries", calculateTimeSeries(bookings, start, end));

                return report;
        }

        private List<Map<String, Object>> calculateTimeSeries(List<Booking> bookings, LocalDateTime start,
                        LocalDateTime end) {
                List<Map<String, Object>> timeSeries = new ArrayList<>();
                DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MMM dd");

                // Generate points for each day in range (limit to reasonable number)
                LocalDateTime current = start;
                while (!current.isAfter(end) && !current.isAfter(LocalDateTime.now())) {
                        LocalDateTime dayStart = current.withHour(0).withMinute(0).withSecond(0);
                        LocalDateTime dayEnd = current.withHour(23).withMinute(59).withSecond(59);
                        String dateLabel = current.format(formatter);

                        long count = bookings.stream()
                                        .filter(b -> b.getCreatedAt() != null && !b.getCreatedAt().isBefore(dayStart)
                                                        && !b.getCreatedAt().isAfter(dayEnd))
                                        .count();

                        double revenue = bookings.stream()
                                        .filter(b -> b.getCreatedAt() != null && !b.getCreatedAt().isBefore(dayStart)
                                                        && !b.getCreatedAt().isAfter(dayEnd))
                                        .filter(b -> b.getPayment() != null && b.getPayment()
                                                        .getStatus() == Booking.PaymentStatus.completed)
                                        .mapToDouble(b -> b.getPayment().getAmount() != null
                                                        ? b.getPayment().getAmount()
                                                        : 0.0)
                                        .sum();

                        Map<String, Object> dataPoint = new HashMap<>();
                        dataPoint.put("date", dateLabel);
                        dataPoint.put("bookings", count);
                        dataPoint.put("revenue", revenue);
                        timeSeries.add(dataPoint);

                        current = current.plusDays(1);
                        if (timeSeries.size() > 60)
                                break; // Safety limit
                }
                return timeSeries;
        }

        private Map<String, Object> calculatePeakHours(List<Booking> bookings) {
                Map<Integer, Integer> hourCounts = new HashMap<>();
                for (Booking b : bookings) {
                        if (b.getStartTime() != null) {
                                int hour = b.getStartTime().getHour();
                                hourCounts.put(hour, hourCounts.getOrDefault(hour, 0) + 1);
                        }
                }

                List<Map<String, Object>> distribution = hourCounts.entrySet().stream()
                                .map(e -> {
                                        Map<String, Object> m = new HashMap<>();
                                        m.put("hour", e.getKey());
                                        m.put("count", e.getValue());
                                        m.put("timeLabel", String.format("%02d:00 - %02d:59", e.getKey(), e.getKey()));
                                        return m;
                                })
                                .sorted((m1, m2) -> (Integer) m2.get("count") - (Integer) m1.get("count"))
                                .collect(Collectors.toList());

                Map<String, Object> peakHours = new HashMap<>();
                peakHours.put("distribution", distribution);
                peakHours.put("peakHour", distribution.isEmpty() ? "N/A" : distribution.get(0).get("timeLabel"));
                return peakHours;
        }

        private Map<String, Object> calculateSegmentation(List<Booking> bookings) {
                Map<String, Object> segmentation = new HashMap<>();

                // By Status
                Map<String, Long> byStatus = bookings.stream()
                                .collect(Collectors.groupingBy(b -> b.getStatus().toString(), Collectors.counting()));
                segmentation.put("byStatus", byStatus);

                // By Slot (List for Recharts)
                List<Map<String, Object>> bySlot = bookings.stream()
                                .filter(b -> b.getSlot() != null)
                                .collect(Collectors.groupingBy(b -> b.getSlot().getSlotNumber(), Collectors.counting()))
                                .entrySet().stream()
                                .map(e -> {
                                        Map<String, Object> m = new HashMap<>();
                                        m.put("slotNumber", e.getKey());
                                        m.put("count", e.getValue());
                                        // Calculate revenue for this slot
                                        double rev = bookings.stream()
                                                        .filter(b -> b.getSlot() != null && e.getKey()
                                                                        .equals(b.getSlot().getSlotNumber()))
                                                        .filter(b -> b.getPayment() != null && b.getPayment()
                                                                        .getStatus() == Booking.PaymentStatus.completed)
                                                        .mapToDouble(b -> b.getPayment().getAmount() != null
                                                                        ? b.getPayment().getAmount()
                                                                        : 0.0)
                                                        .sum();
                                        m.put("revenue", rev);
                                        return m;
                                })
                                .sorted((m1, m2) -> (int) ((long) m2.get("count") - (long) m1.get("count")))
                                .collect(Collectors.toList());
                segmentation.put("bySlot", bySlot);

                // By Location
                List<Map<String, Object>> byLocation = bookings.stream()
                                .filter(b -> b.getSlot() != null && b.getSlot().getAddress() != null)
                                .collect(Collectors.groupingBy(b -> b.getSlot().getAddress(), Collectors.counting()))
                                .entrySet().stream()
                                .map(e -> {
                                        Map<String, Object> m = new HashMap<>();
                                        m.put("address", e.getKey());
                                        m.put("count", e.getValue());
                                        return m;
                                })
                                .sorted((m1, m2) -> (int) ((long) m2.get("count") - (long) m1.get("count")))
                                .collect(Collectors.toList());
                segmentation.put("byLocation", byLocation);

                return segmentation;
        }

        public Map<String, Object> getPersonalStats(String userId) {
                List<Booking> bookings = bookingRepository
                                .findByUserOrderByCreatedAtDesc(
                                                com.parking.validator.model.User.builder().id(userId).build());

                Map<String, Object> stats = new HashMap<>();
                Map<String, Object> summary = new HashMap<>();

                long completedCount = bookings.stream().filter(b -> b.getStatus() == Booking.BookingStatus.COMPLETED)
                                .count();
                double totalSpent = bookings.stream()
                                .filter(b -> b.getPayment() != null
                                                && b.getPayment().getStatus() == Booking.PaymentStatus.completed)
                                .mapToDouble(b -> b.getPayment().getAmount() != null ? b.getPayment().getAmount() : 0.0)
                                .sum();
                long activeCount = bookings.stream()
                                .filter(b -> b.getParkingStatus() == Booking.ParkingStatus.CHECKED_IN)
                                .count();
                double avgDuration = bookings.stream()
                                .filter(b -> b.getActualDuration() != null && b.getActualDuration() > 0)
                                .mapToInt(b -> b.getActualDuration())
                                .average().orElse(0.0);

                // Favorite Location
                String favorite = bookings.stream()
                                .filter(b -> b.getSlot() != null && b.getSlot().getArea() != null)
                                .collect(Collectors.groupingBy(b -> b.getSlot().getArea(), Collectors.counting()))
                                .entrySet().stream()
                                .max(Map.Entry.comparingByValue())
                                .map(Map.Entry::getKey)
                                .orElse("N/A");

                summary.put("totalBookings", bookings.size());
                summary.put("totalSpent", Math.round(totalSpent * 100.0) / 100.0);
                summary.put("activeBookings", activeCount);
                summary.put("completedBookings", completedCount);
                summary.put("averageDuration", Math.round(avgDuration));
                summary.put("favoriteLocation", favorite);

                stats.put("summary", summary);

                // Recent Bookings (Transformed)
                List<Map<String, Object>> recent = bookings.stream().limit(10).map(b -> {
                        Map<String, Object> m = new HashMap<>();
                        m.put("id", b.getId());
                        m.put("slotNumber", b.getSlot() != null ? b.getSlot().getSlotNumber() : "N/A");
                        m.put("date", b.getCreatedAt());
                        m.put("amount", b.getPayment() != null ? b.getPayment().getAmount() : 0.0);
                        m.put("status", b.getStatus().toString());
                        return m;
                }).collect(Collectors.toList());
                stats.put("recentBookings", recent);

                // Time Series (Last 7 days for trend)
                List<Map<String, Object>> trend = new ArrayList<>();
                DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MMM dd");
                for (int i = 6; i >= 0; i--) {
                        LocalDateTime day = LocalDateTime.now().minusDays(i).withHour(0).withMinute(0);
                        String dateLabel = day.format(formatter);

                        long count = bookings.stream()
                                        .filter(b -> b.getCreatedAt() != null &&
                                                        b.getCreatedAt().getYear() == day.getYear() &&
                                                        b.getCreatedAt().getDayOfYear() == day.getDayOfYear())
                                        .count();

                        double revenue = bookings.stream()
                                        .filter(b -> b.getCreatedAt() != null &&
                                                        b.getCreatedAt().getYear() == day.getYear() &&
                                                        b.getCreatedAt().getDayOfYear() == day.getDayOfYear())
                                        .filter(b -> b.getPayment() != null && b.getPayment()
                                                        .getStatus() == Booking.PaymentStatus.completed)
                                        .mapToDouble(b -> b.getPayment().getAmount() != null
                                                        ? b.getPayment().getAmount()
                                                        : 0.0)
                                        .sum();

                        Map<String, Object> dataPoint = new HashMap<>();
                        dataPoint.put("date", dateLabel);
                        dataPoint.put("bookings", count);
                        dataPoint.put("revenue", revenue);
                        trend.add(dataPoint);
                }
                stats.put("timeSeries", trend);

                // Peak Hours (Personal)
                Map<String, Object> personalPeak = calculatePeakHours(bookings);
                stats.put("peakHours", personalPeak.get("distribution"));

                return stats;
        }

        public String generateCSV(List<Booking> bookings) {
                StringBuilder csv = new StringBuilder();
                csv.append("Booking ID,Date,User Name,Slot Number,Vehicle Number,Start Time,End Time,Amount,Status\n");

                DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");

                for (Booking b : bookings) {
                        csv.append(String.format("%s,%s,%s,%s,%s,%s,%s,%.2f,%s\n",
                                        b.getId(),
                                        b.getCreatedAt() != null ? b.getCreatedAt().format(formatter) : "N/A",
                                        b.getUser() != null ? b.getUser().getName() : "N/A",
                                        b.getSlot() != null ? b.getSlot().getSlotNumber() : "N/A",
                                        b.getVehicleNumber(),
                                        b.getStartTime() != null ? b.getStartTime().format(formatter) : "N/A",
                                        b.getEndTime() != null ? b.getEndTime().format(formatter) : "N/A",
                                        b.getPayment() != null ? b.getPayment().getAmount() : 0.0,
                                        b.getStatus()));
                }
                return csv.toString();
        }
}
